internal/api/watch_promotion_v1alpha1.go

```diff
@@ -60,17 +60,17 @@ func (s *server) WatchPromotion(
 			}
 			u, ok := e.Object.(*unstructured.Unstructured)
 			if !ok {
-				return errors.Errorf("unexpected object type %T", e.Object)
+				return connect.NewError(connect.CodeInternal, errors.Errorf("unexpected object type %T", e.Object))
```

Comment:
Is this a place where you meant to return a raw error and missed it?

Reply:
No, since these errors are explicit `internal` errors, I changed to return `internal` status code instead of returning raw error.

Reply:
I'm not sure I understand. Isn't the handler's whole purpose to make sense of whatever error is returned to it?

* If it's a connect error, no interpretation is necessary.
* If it's a k8s error, map it to a connect error.
* If it's anything else, it was completely unanticipated and therefore amounts, by default, to an internal service error.

Am I missing something?

Reply:
I want to make things clear - I think the `handler`'s purpose is to handle request and return appropriate _response_ with explicit status code to the client.

From that perspective, if the handler returns non-connect error, the error should be treated as some kind of _RuntimeException_ - so the status code should be `unknown` instead of `internal` since the handler doesn't expected such error. In this case, we tried to cast object's type but failed - the handler expected this error so the error code should be `internal`, not `unknown`.

Reply:
\+ https://github.com/akuity/kargo/pull/1266#issuecomment-1874249046

Yes, handlers can be tested without interceptor. However, IMHO, returning raw k8s error and mapping error codes in interceptor will cause uncertainty of API responses since the handler doesn't know how the status code will be mapped to. 

> I think the handler's purpose is to handle request and return appropriate response with explicit status code to the client.

Which makes us not able to check response's status code. We can assert error `type` in tests for sure, but I believe it will make hard to write test code since it should deeply know the implementation of called functions/methods in handler function, not the abstracted error.

Reply:
Ok... I think there are a few different things going on here that we need to separate and deal with more or less independently:

1. > I think the handler's purpose is to handle request and return appropriate response with explicit status code to the client.

    It makes total sense for handlers to be explicit with the error codes they return _in certain cases where specific failures are anticipated_. As an example, if validation of a request failed, it would make sense for a handler to explicitly return something like "bad request" (I think connect calls it "invalid argument").
    
    In the general case -- i.e. an unanticipated error -- it is onerous for developers to explicitly specify an error code. Requiring them to do so also creates lots of opportunity for human error. _It is too easy to forget._

2. If you accept no. 1, then the role of the interceptor is to try to make sense of unanticipated errors -- which most errors are likely to be.

    > if the handler returns non-connect error, the error should be treated as some kind of RuntimeException
    
    Agree.
    
    > so the status code should be unknown instead of internal since the handler doesn't expected such error.

    This is where we seem to disagree. "I didn't expect this to happen" is different from "I don't know what happened" and "internal server error" has always been the conventional way of communicating to a client that the failure was something unexpected on the server's end. Can you explain why this should be "unknown" instead?

3. I think since there are places in handlers where you are explicitly choosing  "internal server error," you must also perceive there to be some difference between "internal server error" and "unknown," but its not clear to me what's different between the spots where you are explicit and where you are not.

4. I think the interceptor is not currently handling the case of an error that _wasn't_ a Kubernetes error or a connect error. It should, and per no.2, I believe such cases should default to "internal server error."

5. I don't understand why the tests for the interceptor are so complex and standing up an HTTP server and using dummy handlers. The logic that maps errors to to connect errors could be tested by directly invoking the `wrapError()` function. I think the test is also missing many cases -- such as the errors mentioned in no.4 and also errors that have been wrapped using `errors.Wrap()`.

6. Sort of separate from the rest of it:

    > Yes, handlers can be tested without interceptor. However, IMHO, returning raw k8s error and mapping error codes in interceptor will cause uncertainty of API responses since the handler doesn't know how the status code will be mapped to.

    Returning to no.1, interpretation of errors isn't a handler's task.
    
    In general, I'd like us to embrace true _unit_ testing where we can. i.e. Testing components in isolation. Test cases simulate the layer "above" that is (directly, whenever possible) invoking the component/function under test, with dependencies _of_ the component/function under test mocked out (wherever practical). Integration tests have their place, and I am certain that we will eventually build a suite of e2e tests, but the distinct advantage of testing components in isolation, to the extent possible, is that it tremendously narrows down the source of problems and makes it easier to exercise edge cases.

Despite the length of this comment... I think this PR is very close to where it needs to be. I think we just need to come to some agreement on a few nitty gritty details and then things will fall into place.